File: src/utils/JwtUtil.java
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.util.Date;

@Component
public class JwtUtil {

    @Value("${jwt.secret}")
    private String secretKey;

    @Value("${jwt.expiration}")
    private long expirationTime;

    /**
     * Gera um token JWT usando o nome de usuário e define uma data de expiração.
     *
     * @param username o nome de usuário para o qual o token é gerado
     * @return o token JWT gerado como uma String
     */
    public String generateToken(String username) {
        return Jwts.builder()
                .setSubject(username)
                .setIssuedAt(new Date())
                .setExpiration(new Date(System.currentTimeMillis() + expirationTime))
                .signWith(SignatureAlgorithm.HS256, secretKey)
                .compact();
    }

    /**
     * Valida um token JWT, verificando sua assinatura e expiração.
     *
     * @param token o token JWT a ser validado
     * @return true se o token for válido, caso contrário false
     */
    public boolean validateToken(String token) {
        try {
            Jwts.parser().setSigningKey(secretKey).parseClaimsJws(token);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Extrai o nome de usuário de um token JWT.
     *
     * @param token o token JWT do qual extrair o nome de usuário
     * @return o nome de usuário contido no token
     */
    public String getUsernameFromToken(String token) {
        Claims claims = Jwts.parser()
                .setSigningKey(secretKey)
                .parseClaimsJws(token)
                .getBody();
        return claims.getSubject();
    }

    /**
     * Extrai as "claims" (informações adicionais) de um token JWT.
     *
     * @param token o token JWT do qual extrair as claims
     * @return as claims contidas no token
     */
    public Claims getClaimsFromToken(String token) {
        return Jwts.parser()
                .setSigningKey(secretKey)
                .parseClaimsJws(token)
                .getBody();
    }
}

File: src/utils/JwtUtil.js
const jwt = require('jsonwebtoken');


const SECRET_KEY = process.env.JWT_SECRET || 'my_secret_key';  // Recomendado definir em variáveis de ambiente
const EXPIRATION_TIME = process.env.JWT_EXPIRATION || '1h';    // 1 hora (use '1h', '2d', etc.)

/**
 * Gera um token JWT usando o nome de usuário e define uma data de expiração.
 *
 * @param {string} username - Nome de usuário para o qual o token é gerado
 * @returns {string} - Token JWT gerado
 */
function generateToken(username) {
    return jwt.sign({ username }, SECRET_KEY, { expiresIn: EXPIRATION_TIME });
}

/**
 * Valida um token JWT verificando sua assinatura e expiração.
 *
 * @param {string} token - Token JWT a ser validado
 * @returns {boolean} - Retorna true se o token for válido, caso contrário false
 */
function validateToken(token) {
    try {
        jwt.verify(token, SECRET_KEY);
        return true;
    } catch (error) {
        return false;
    }
}

/**
 * Extrai o nome de usuário de um token JWT.
 *
 * @param {string} token - Token JWT do qual extrair o nome de usuário
 * @returns {string|null} - Retorna o nome de usuário ou null se o token for inválido
 */
function getUsernameFromToken(token) {
    try {
        const decoded = jwt.verify(token, SECRET_KEY);
        return decoded.username;
    } catch (error) {
        return null;
    }
}

/**
 * Extrai as "claims" (informações adicionais) de um token JWT.
 *
 * @param {string} token - Token JWT do qual extrair as claims
 * @returns {object|null} - Retorna as claims contidas no token ou null se o token for inválido
 */
function getClaimsFromToken(token) {
    try {
        return jwt.verify(token, SECRET_KEY);
    } catch (error) {
        return null;
    }
}

// Exporta as funções para uso em outros módulos
module.exports = {
    generateToken,
    validateToken,
    getUsernameFromToken,
    getClaimsFromToken,
};
